Backend/manual parsing

nlahlborg opened the pull request:
# Pull Request

## Description
manual parsing for 3 common job site formats
- LinkedIn EasyApply
- @us.greenhouse-mail.io
- @hire.lever.co

## Related Issue
related to [issue#212](https://github.com/lnovitz/jobseeker-analytics/issues/212)

## Type of Change
<!-- Put an x in the boxes that apply -->
- [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
- [ ] ✨ New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📝 Documentation update
- [ ] 🎨 Style update (formatting, renaming)
- [ ] ♻️ Code refactor (no functional changes)
- [x] ⚡️ Performance improvement
- [ ] ✅ Test update
- [ ] 🔨 Build configuration update
- [ ] 🔒 Security update

## Changes Made
<!-- List the changes made in this PR -->
1. moved some parsing functions from test_filter_schema to filter_utils
2. coded the apply_override_filter logic
3. created test cases
4. created yaml config for emails that should have manual parsing
5. created 3 hardcoded manual parsers in email_utils
6. created logic to route messages to manual parser or llm

## Testing
<!-- Describe the tests you ran and how to reproduce them -->
1. Steps to reproduce:
   - pytest --> all unit tests should pass
   - run the entire app flow in browser
 
2. Test environment:
   - OS: Docker container

## Checklist
<!-- Put an x in the boxes that apply -->
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

## Performance Impact
<!-- Describe any performance impacts this change may have -->
- [ ] No performance impact
- [x] Performance impact (describe below):
  - expect faster output of results for users with a lot of LinkedIn EasyApply applications

## Security Considerations
<!-- Describe any security implications this change may have -->
- [x] No security implications
- [ ] Security implications considered (describe below):
  - 

## Additional Notes
there seems to be something broken with local implementation of the app on the production branch for me... even after processing is finished, my csv download is not immediately available. 


---

By submitting this pull request, I confirm that my contribution is made under the terms of the project's license.

lnovitz commented on backend/utils/email_utils.py:
```diff
@@ -384,3 +387,41 @@ def get_top_consecutive_capitalized_words(tuples_list):
+    result = None
+    if apply_override_filter(subject_text=msg["subject"], from_text=msg["from"], filter_config=filter_config):
+        result = process_email_manual(msg)
+        processor = "manual"
```
Smart

lnovitz reviewed (approved):
I really need to look at this more but I tested and it works and it formats the company name better than before so sure

nlahlborg commented:
> I really need to look at this more but I tested and it works and it formats the company name better than before so sure

I'm sorry it's sooooo long. I recycled some functions I wrote previously only for unit tests and moved them to a different part of the code accordingly... so that added a lot of lines. But you already commented on the actual "meat" of this change.

I would like to draw your attention to [this part](https://github.com/lnovitz/jobseeker-analytics/blob/a2b486b5002ce3d69ccd057ada3adbb417398ba0/backend/utils/email_utils.py#L401), where I've added in this extra filter so that I can do unit testing without actually needing to create a new email test case. Maybe you have some thoughts?

lnovitz reviewed (changes requested):
So I honestly hate my use of `SAMPLE_MESSAGE` even though I'm emotionally attached to it because I spent probably about an hour redacting potentially sensitive email information from an actual email to create it. But anyway, it makes unit tests hard to read.

I ended up trying to modify your `applied_email_filter_manual_parse.yaml` but I found it was a little too restrictive. There's a lot of ways we can use company name from the subject line. So I'm thinking we can just use regex as much as possible regardless of the subject / domain match.

The only thing I'm not sure of is the `application_status`. I think for the most part, the regex I captured 
```        
r"Your application was sent to ([\w\s]+)",
        r".*(?:applying|applied) (?:to|at) ([\w\s]+)",
        r".*Application to ([\w\s]+)",
        r".*Interview with ([\w\s]+)",
        r".*for your interest in ([A-Za-z]+)",
        r".*about your application to ([\w\s]+)",
        r".*received your ([\w\s]+) application*",
        r".*Thank you from ([\w\s]+)",
```
is easy enough to filter emails down to a rejection or a "no response" (no response meaning just a confirmation of application)

so I think if the company name is successfully extracted manually from any of the regex patterns in `extract_company_name_from_subject`, we can then do an email body content regex match for rejection type words. If none found, return "no response". (or alternately you can try `SentenceTransformer` technique here https://github.com/lnovitz/jobseeker-analytics/blob/f62b6f5a397c314a4fe1706b226fdd07014bfa11/backend/utils/minilm_utils.py) What do you think?

nlahlborg commented:
ok I was deliberately trying to be very rigid and narrow with the scope because I'm a scaredy cat about deploying big code changes onto production. The conservative approach results in successful manual processing of ~50% of emails. If you want to instead be more aggressive and target something like 90% transition to manual parsing then I definitely like your approach. I can take a more detailed look at the implementation and do some debug/test tonight. Just confirm that you want such a big change, since you're the maintainer!

lnovitz commented:
Ideally we don't use third party APIs like Gemini. That's the best state. Do everything in house for performance and speed as well as users would probably like that we aren't sending their emails to LLMs :) never mind that it's very job related specific

lnovitz commented:
So yes I'm all for manual parsing. Simple is better. But if sentence embedding model is faster than adding regex for everything + corresponding unit tests, then we should try some kind of language model. I looked into MiniLM, since it's small
